### PR TITLE
Notes about Fastlane Match.

### PR DESCRIPTION
**Pull request by zzcgumn**
Last week I had to do work on creating a version of the Babylon app that is signed for ad hoc distribution. It took me a fair amount of time to figure out how this can be done with Fastlane Match. The information I needed was scattered in several places. Hopefully you will be lucky enough that running `bundle exec fastlane match` is all you will ever need to do. If you are unlucky, I hope that this document will save you some time.

### CODE REVIEW

File: TechnicalDocuments/FastlaneMatch.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+```shell
+irb(main):007:0> commit_message = 'Updated provisioning profiles to be great again.'
+=> "Updated provisioning profiles to be great again.”
+irb(main):008:> Match::GitHelper.commit_changes(workspace, commit_message, git_url, branch)
+[14:28:44]: 🔒  Successfully encrypted certificates repo
+[14:28:44]: Pushing changes to remote git repo…
+=> nil
+```
+
+It can be convenient to store the password in the environment hash
```
Comment (ilyapuchka) [inline TechnicalDocuments/FastlaneMatch.md]: We can add: "The password itself can be found in the iOS team 1Password vault"

File: TechnicalDocuments/FastlaneMatch.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ certs/development
+ certs/distribution
+ certs/enterprise
+```
+
+It might be necessary to manually make a certificate available in both the enterprise and distribution folder. Copying the files did not work (February 2019), but adding a soft link did.
+
+Fastlane Math expects the filename for the .p12 and .cer to be `certificate id`.p12 and `certificate id`.cer, whereas manually generated certificates that have been downloaded from the developer portal are named `team id`.cer and `team id`.p12.
+
+Fastlane has a utility for interacting with Apple Developer Portal and the App Store Connect API called [spaceship](https://github.com/fastlane/fastlane/tree/master/spaceship). There were some problems (March 2019) with running bundle exec fastlane spaceship. The certificate id can be found by inspecting the certificate in the spaceship playground.
```
Comment (ilyapuchka) [inline TechnicalDocuments/FastlaneMatch.md]:
```suggestion
Fastlane has a utility for interacting with Apple Developer Portal and the App Store Connect API called [spaceship](https://github.com/fastlane/fastlane/tree/master/spaceship). There were some problems (March 2019) with running `bundle exec fastlane spaceship`. The certificate id can be found by inspecting the certificate in the spaceship playground.
```

File: TechnicalDocuments/FastlaneMatch.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
+Fastlane Match
+==============
+
+Fastlane Match is a tool for managing certificates and provisioning profiles centrally. It downloads certificates and profiles from a GitHub repository or Google Cloud Storage bucket. For safety centrally stored files are encrypted.
+
+Fastlane match [documentation](https://docs.fastlane.tools/actions/match/).
+
+## Several Targets and Types
```
Comment (ilyapuchka) [inline TechnicalDocuments/FastlaneMatch.md]: I guess it will make sense to add a section about general usage of `fastlane match`, like what you need to do to get certificates locally, essentially running it in readonly mode with correct parameters. And how certificates are stored right now (i.e. that appstore certificates are in separate branch by the name of appstore org)

File: TechnicalDocuments/FastlaneMatch.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+Best practise for managing different bundle ids is to create one branch for each of them. Default values for the git URL, branch, app identifier and build type are read from the `Matchfile`.
+
+It can happen that code signing certificates for more than one target or type is needed for a build. This can be handled by making several calls to match.
+
+```ruby
+match(app_identifier: "com.best.app.ever.seriously", type: "adhoc", git_branch: "best_app_ever_ad_hoc_signing")
+match(app_identifier: "com.best.app.ever.seriously", type: "development", git_branch: "best_app_ever_dev_signing")
+```
+
+Information about which repository and branches we use for the Babylon apps can be found in our internal Confluence Wiki.
```
Comment (daniloaliberti) [inline TechnicalDocuments/FastlaneMatch.md]: Information about which repository and branches we use for the Babylon apps can be found on the [new-hires-checklist](https://github.com/Babylonpartners/ios-playbook#4-new-hires-checklist).

File: TechnicalDocuments/FastlaneMatch.md (hunk 3 above)
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: 👍 

I will add a section about this tomorrow.

File: TechnicalDocuments/FastlaneMatch.md (hunk 4 above)
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: 👍

File: TechnicalDocuments/FastlaneMatch.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+## Update Contents Manually
+
+Fastlane Match provides a utility for nuking all existing certificates and provisioning profiles and replace them with new ones generated by Fastlane Match. This is, however, not always possible. Apple does not allow more than two distribution certificates per account, so it can happen that Fastlane Match cannot create a new distribution certificate and is unable to figure out which existing certificate to link with the provisioning profile. Nuking existing enterprise certificates can be a really bad idea as in house apps will stop working.
+
+Since the files in the GitHub repository are encrypted one cannot update them directly. Fastlane Match provides a couple of helper functions. To clone the repository, open a ruby console and call `Match::GitHelper.clone`. Babylon specific credentials can be found in our 1Password account.
```
Comment (daniloaliberti) [inline TechnicalDocuments/FastlaneMatch.md]: Since the files in the GitHub repository are encrypted one**s, we** cannot update them directly.

File: TechnicalDocuments/FastlaneMatch.md (hunk 1 above)
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: 👍

File: TechnicalDocuments/FastlaneMatch.md (hunk 2 above)
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: 👍

File: TechnicalDocuments/FastlaneMatch.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
+Fastlane Match
+==============
+
+Fastlane Match is a tool for managing certificates and provisioning profiles centrally. It downloads certificates and profiles from a GitHub repository or Google Cloud Storage bucket. For safety centrally stored files are encrypted.
```
Comment (mluisbrown) [inline TechnicalDocuments/FastlaneMatch.md]:
```suggestion
Fastlane Match is a tool for managing certificates and provisioning profiles centrally. It downloads certificates and profiles from a GitHub repository or Google Cloud Storage bucket. For safety, centrally stored files are encrypted.
```

File: TechnicalDocuments/FastlaneMatch.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+Fastlane match [documentation](https://docs.fastlane.tools/actions/match/).
+
+## Several Targets and Types
+
+Best practise for managing different bundle ids is to create one branch for each of them. Default values for the git URL, branch, app identifier and build type are read from the `Matchfile`.
+
+It can happen that code signing certificates for more than one target or type is needed for a build. This can be handled by making several calls to match.
```
Comment (mluisbrown) [inline TechnicalDocuments/FastlaneMatch.md]:
```suggestion
It can happen that code signing certificates for more than one target or type are needed for a build. This can be handled by making several calls to match.
```

File: TechnicalDocuments/FastlaneMatch.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+ certs/development
+ certs/distribution
+ certs/enterprise
+```
+
+It might be necessary to manually make a certificate available in both the enterprise and distribution folder. Copying the files did not work (February 2019), but adding a soft link did.
+
+Fastlane Math expects the filename for the .p12 and .cer to be `certificate id`.p12 and `certificate id`.cer, whereas manually generated certificates that have been downloaded from the developer portal are named `team id`.cer and `team id`.p12.
+
+Fastlane has a utility for interacting with Apple Developer Portal and the App Store Connect API called [spaceship](https://github.com/fastlane/fastlane/tree/master/spaceship). There were some problems (March 2019) with running `bundle exec fastlane spaceship`. It might be necessary to manually install some gems in the `~/.fastlane` directory.
```
Comment (mr-v) [inline TechnicalDocuments/FastlaneMatch.md]: Please consider adding another title to make this section stand out more, e.g.

```
## Accessing Apple Developer Portal with fastlane spaceship
```

Comment (mr-v) [reply TechnicalDocuments/FastlaneMatch.md]: I checked now and for me only missing `gem` was `pry`. Is that the issue you encountered?

If yes we could add `pry` to our `Gemfile` and remove this section:

> There were some problems (March 2019) with running `bundle exec fastlane spaceship`. It might be necessary to manually install some gems in the `~/.fastlane` directory.

My error logs after running `bundle exec fastlane spaceship`:
```console
Could not find gem 'pry'

If you installed spaceship using `sudo gem install spaceship` run
  sudo gem install pry
to install the missing gem

If you use a Gemfile add this to your Gemfile:
  gem 'pry'
and run `bundle install`
```

File: TechnicalDocuments/FastlaneMatch.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
+Fastlane Match
+==============
+
+Fastlane Match is a tool for managing certificates and provisioning profiles centrally. It downloads certificates and profiles from a GitHub repository or Google Cloud Storage bucket. For safety centrally stored files are encrypted.
+
+Fastlane match [documentation](https://docs.fastlane.tools/actions/match/).
+
+## Several Targets and Types
+
+Best practise for managing different bundle ids is to create one branch for each of them. Default values for the git URL, branch, app identifier and build type are read from the `Matchfile`.
```
[the remark this answers is not in the document]
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: Good spot, I did intend to try to say that there should be in branch per team. Thanks.

File: TechnicalDocuments/FastlaneMatch.md (hunk 8 above)
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: Unfortunately, adding `pry` to my Gemfile gave version errors. Have you tested if it works for you @mr-v ?

File: TechnicalDocuments/FastlaneMatch.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+```ruby
+match(app_identifier: "com.best.app.ever.seriously", type: "adhoc", git_branch: "best_app_ever_ad_hoc_signing")
+match(app_identifier: "com.best.app.ever.seriously", type: "development", git_branch: "best_app_ever_dev_signing")
+```
+
+Repository URLs can be found in the [new hires checklist](https://github.com/Babylonpartners/ios-playbook#4-new-hires-checklist). Information about which branches we use for different Babylon apps can be found in our internal Confluence Wiki.
+
+## Update Contents Manually
+
+Fastlane Match provides a utility for nuking all existing certificates and provisioning profiles and replace them with new ones generated by Fastlane Match. This is, however, not always possible. Apple does not allow more than two distribution certificates per account, so it can happen that Fastlane Match cannot create a new distribution certificate and is unable to figure out which existing certificate to link with the provisioning profile. Nuking existing enterprise certificates can be a really bad idea as in house apps will stop working.
```
[the remark this answers is not in the document]
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: One reason it does not always work is that we cannot freely revoke and replace certificates that are used in enterprise distribution provisioning profiles. In house apps that have been signed with them stops working which upsets our sales and marketing teams.

File: TechnicalDocuments/FastlaneMatch.md (hunk 8 above)
Comment (mr-v) [reply TechnicalDocuments/FastlaneMatch.md]: Yes, I can login.

File: TechnicalDocuments/FastlaneMatch.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: I will to try rework this paragraph to make it clear that it is not always possible for fastlane match to create certificates without nuking existing ones, but nuking existing certificates can be a really bad idea.

File: TechnicalDocuments/FastlaneMatch.md (hunk 8 above)
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: I can login after installing `pry` manually, but adding `pry` to the Gemfile did not work.

File: TechnicalDocuments/FastlaneMatch.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+==============
+
+Fastlane Match is a tool for managing certificates and provisioning profiles centrally. It downloads certificates and profiles from a GitHub repository or Google Cloud Storage bucket. For safety, centrally stored files are encrypted.
+
+## Installing Certificates and Provisioning Profiles Locally
+
+For day to day development work it usually suffices to navigate to the project directory in a terminal shell and run the command
+
+```shell
+ $ fastlane match development --read-only
```
Comment (mr-v) [inline TechnicalDocuments/FastlaneMatch.md]:
```suggestion
 $ fastlane match development --readonly
```

File: TechnicalDocuments/FastlaneMatch.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+```
+
+Certificates and provisioning profiles for enterprise builds (Hockey App) or release builds (Testflight and the App Store) can also be downloaded. Normally, this is not needed but might be necessary if you want to run UI tests locally or work on updating our fastlane build scripts.
+
+To install certificates and provisioning profiles for creating a Testflight or App Store build locally, run the command
+
+```shell
+ $ fastlane match release git_branch:"release team id" --read-only
```
Comment (mr-v) [inline TechnicalDocuments/FastlaneMatch.md]:
```suggestion
 $ fastlane match release git_branch:"release team id" --readonly
```

File: TechnicalDocuments/FastlaneMatch.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+```
+
+For a Hockeyapp build
+
+```shell
+ $ fastlane match enterprise --read-only
```
Comment (mr-v) [inline TechnicalDocuments/FastlaneMatch.md]:
```suggestion
 $ fastlane match enterprise --readonly
```

File: TechnicalDocuments/FastlaneMatch.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+=> "long-and-super-secret"
+irb(main):006:0> workspace = Match::GitHelper.clone(git_url, shallow_clone, manual_password: password, branch: branch)
+[14:22:10]: Cloning remote git repo…
+[14:22:10]: If cloning the repo takes too long, you can use the `clone_branch_directly` option in match.
+[14:22:12]: Checking out branch build_target…
+[14:22:12]: 🔓  Successfully decrypted certificates repo
+=> "/var/folders/8k/r0x0ys_927q8vq5_tq01ntjd2b03m3/T/d20190228-92193-16w4fz2"
+```
+
+Open another terminal session and navigate to the folder name returned by `Match::GitHelper.clone`. Once all edits are complete, return to the ruby console.
```
[the remark this answers is not in the document]
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: 👍

File: TechnicalDocuments/FastlaneMatch.md (hunk 8 above)
Comment (AliSoftware) [reply TechnicalDocuments/FastlaneMatch.md]: @zzcgumn Are you using zsh, especially oh-my-zsh with the bundler plugin? If so, it automatically wraps calls to some gems into `bundle exec`, and `pry` is one of them by default. Meaning that running `pry` in such a zsh shell will actually run `bundle exec pry`, then seing that you don't have `pry` in your `Gemfile`, will abort... maybe that's related?

(FYI In such cases you can use `unbundled_pry` to run the raw `pry` command without zsh wrapping it in `bundle exec` for you, or you can edit your `.zshrc` to add `UNBUNDLED_COMMANDS=(pry)` before the `plugins=...` line)

File: TechnicalDocuments/FastlaneMatch.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+```shell
+ $ fastlane match enterprise --read-only
+```
+
+More information about available options can be found below and in the fastlane match [documentation](https://docs.fastlane.tools/actions/match/).
+
+## Several Targets and Teams
+
+Recommended practise is to have on branch per team. It is also possible to create branches for different bundle identifiers should it get messy with a lot of app identifiers for one team. Default values for the git URL, branch, app identifier and build type are read from the `Matchfile`.
```
[the remark this answers is not in the document]
Comment (AliSoftware) [reply TechnicalDocuments/FastlaneMatch.md]: Agreed, this portion might be confusing.

Also `s/practise/practice/`

File: TechnicalDocuments/FastlaneMatch.md (hunk 5 above)
Comment (AliSoftware) [reply TechnicalDocuments/FastlaneMatch.md]: I think the meaning was "Since the files in the GitHub repository are encrypted, one cannot update them directly" 😛 ("One" used here as the indefinite pronoun). but agree a comma was missing nonetheless 😉

Comment (daniloaliberti) [reply TechnicalDocuments/FastlaneMatch.md]: ah `one` as 'someone' 👍 got it.

File: TechnicalDocuments/FastlaneMatch.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+## Installing Certificates and Provisioning Profiles Locally
+
+For day to day development work it usually suffices to navigate to the project directory in a terminal shell and run the command
+
+```shell
+ $ fastlane match development --readonly
+```
+
+Certificates and provisioning profiles for enterprise builds (Hockey App) or release builds (Testflight and the App Store) can also be downloaded. Normally, this is not needed but might be necessary if you want to run UI tests locally or work on updating our fastlane build scripts.
```
Comment (ilyapuchka) [inline TechnicalDocuments/FastlaneMatch.md]: I don't think this is really needed for running UI tests, but it is definitely needed for making manual builds.

```suggestion
Certificates and provisioning profiles for enterprise builds (Hockey App) or release builds (Testflight and the App Store) can also be downloaded. Normally, this is not needed but might be necessary if you want to make a release build locally or work on updating our fastlane build scripts.
```

File: TechnicalDocuments/FastlaneMatch.md (hunk 12 above)
Comment (ilyapuchka) [reply TechnicalDocuments/FastlaneMatch.md]: It must be `enterprise` for HockeyApp builds (which are using our enterprise account) or `appstore` for TestFlight/AppStore builds

```suggestion
 $ fastlane match appstore git_branch:"appstore team id" --read-only
```

File: TechnicalDocuments/FastlaneMatch.md [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+```shell
+ $ fastlane match enterprise --readonly
+```
+
+More information about available options can be found below and in the fastlane match [documentation](https://docs.fastlane.tools/actions/match/).
+
+## Several Targets and Teams
+
+Recommended practise is to have on branch per team. It is also possible to create branches for different bundle identifiers should it get messy with a lot of app identifiers for one team. Default values for the git URL, branch, app identifier and build type are read from the `Matchfile`.
```
Comment (ilyapuchka) [inline TechnicalDocuments/FastlaneMatch.md]:
```suggestion
Recommended practise is to have one branch per team. It is also possible to create branches for different bundle identifiers should it get messy with a lot of app identifiers for one team. Default values for the git URL, branch, app identifier and build type are read from the `Matchfile`.
```

File: TechnicalDocuments/FastlaneMatch.md [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+```ruby
+match(app_identifier: "com.best.app.ever.seriously", type: "adhoc", git_branch: "best_app_ever_ad_hoc_signing")
+match(app_identifier: "com.best.app.ever.seriously", type: "development", git_branch: "best_app_ever_dev_signing")
+```
+
+Repository URLs can be found in the [new hires checklist](https://github.com/Babylonpartners/ios-playbook#4-new-hires-checklist). Information about which branches we use for different Babylon apps can be found in our internal Confluence Wiki.
+
+## Update Contents Manually
+
+Apple does not allow more than two distribution certificates per account, so it can happen that fastlane match cannot create a new distribution certificate and is unable to figure out which existing certificate to link with a new provisioning profile. Fastlane match provides a utility for nuking all existing certificates and provisioning profiles and replace them with new ones generated by fastlane match. This is, however, not always advisable. Nuking existing enterprise certificates can be a really bad idea as in house apps will stop working.
```
Comment (ilyapuchka) [inline TechnicalDocuments/FastlaneMatch.md]:
```suggestion
Apple does not allow more than two distribution certificates per account, so it can happen that fastlane match cannot create a new distribution certificate and is unable to figure out which existing certificate to link with a new provisioning profile. Fastlane match provides a utility to revoke and delete all existing certificates and provisioning profiles and replace them with new ones generated by fastlane match. This is, however, not always advisable. Revoking existing enterprise certificates will make all existing in-house builds (that said all existing HockeyApp builds) stop working. It does not affect any existing AppStore builds submitted to TestFlight or AppStore in any way.
```

Comment (ilyapuchka) [inline TechnicalDocuments/FastlaneMatch.md]: It shouldn't really happen that we need more than one distribution certificate. I'd rephrase it and maybe would hide the whole section under `summary` tag as this is something we wouldn't ever need to do any more since we now fully manage our certificates via match.

File: TechnicalDocuments/FastlaneMatch.md (hunk 8 above)
Comment (ilyapuchka) [reply TechnicalDocuments/FastlaneMatch.md]: >  I'd keep instructions simple and only add problems note and workaround information when it becomes a more widespread issue.

I agree with @mr-v on that. As these instructions are far from what we need to deal with on a regular basis, I'd either move it to another document about troubleshooting signing issues or would hide it under `summary` as I suggested before.

File: TechnicalDocuments/FastlaneMatch.md [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+
+Certificates are stored in three folders
+
+```shell
+ certs/development
+ certs/distribution
+ certs/enterprise
+```
+
+It might be necessary to manually make a certificate available in both the enterprise and distribution folder. Copying the files did not work (February 2019), but adding a soft link did.
```
Comment (ilyapuchka) [inline TechnicalDocuments/FastlaneMatch.md]: This is very tricky advice. Enterprise certificates are only available for enterprise teams, distribution (AppStore) - only for "regular" teams. Why and how would one use the same certificate for enterprise and AppStore builds then?

File: TechnicalDocuments/FastlaneMatch.md [hunk 20]
```diff
@@ -0,0 +1,121 @@
+Fastlane Match
```
[the remark this answers is not in the document]
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: I think this should be capitalised as it is the main title.

File: TechnicalDocuments/FastlaneMatch.md (hunk 19 above)
Comment (zzcgumn) [reply TechnicalDocuments/FastlaneMatch.md]: Ad hoc builds.
